libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -33,6 +31,244 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+  _Type __desired = __op(__expected);
+  while (!__cuda_atomic_compare_exchange(__ptr, __expected, __expected, __desired, _Order{}, _Operand{}, _Sco{}))
+  {
+    __desired = __op(__expected);
+  }
+  return __expected;
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void
+__cuda_atomic_store(_Type* __ptr, _Type __val, _Order, _Operand, _Sco, __atomic_cuda_mmio_disable)
+{
+  // Store requires cas on 8/16b types
+  __cuda_atomic_fetch_update(
+    __ptr,
+    [__val](_Type __old) {
+      return __val;
+    },
+    _Order{},
+    __atomic_cuda_operand_tag<__atomic_cuda_operand::_b, _Operand::__size>{},
+    _Sco{});
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_arithmetic<_Operand> = 0>
+static inline _CCCL_DEVICE void __cuda_atomic_fetch_add(_Type* __ptr, _Type& __dst, _Type __op, _Order, _Operand, _Sco)
+{
+  __dst = __cuda_atomic_fetch_update(
+    __ptr,
+    [__op](_Type __old) {
+      return __old + __op;
+    },
+    _Order{},
+    __atomic_cuda_operand_tag<__atomic_cuda_operand::_b, _Operand::__size>{},
+    _Sco{});
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void __cuda_atomic_fetch_and(_Type* __ptr, _Type& __dst, _Type __op, _Order, _Operand, _Sco)
+{
+  __dst = __cuda_atomic_fetch_update(
+    __ptr,
+    [__op](_Type __old) {
+      return __old & __op;
+    },
+    _Order{},
+    __atomic_cuda_operand_tag<__atomic_cuda_operand::_b, _Operand::__size>{},
+    _Sco{});
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void __cuda_atomic_fetch_xor(_Type* __ptr, _Type& __dst, _Type __op, _Order, _Operand, _Sco)
+{
+  __dst = __cuda_atomic_fetch_update(
+    __ptr,
+    [__op](_Type __old) {
+      return __old ^ __op;
+    },
+    _Order{},
+    __atomic_cuda_operand_tag<__atomic_cuda_operand::_b, _Operand::__size>{},
+    _Sco{});
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void __cuda_atomic_fetch_or(_Type* __ptr, _Type& __dst, _Type __op, _Order, _Operand, _Sco)
+{
+  __dst = __cuda_atomic_fetch_update(
+    __ptr,
+    [__op](_Type __old) {
+      return __old | __op;
+    },
+    _Order{},
+    __atomic_cuda_operand_tag<__atomic_cuda_operand::_b, _Operand::__size>{},
+    _Sco{});
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_arithmetic<_Operand> = 0>
+static inline _CCCL_DEVICE void __cuda_atomic_fetch_min(_Type* __ptr, _Type& __dst, _Type __op, _Order, _Operand, _Sco)
+{
+  __dst = __cuda_atomic_fetch_update(
+    __ptr,
+    [__op](_Type __old) {
+      return __old < __op ? __old : __op;
```

Comment:
_I think_ that, in general (not sure what `_Type`s we support here as an extension):
-  `fetch_min(op)` should be: `op < old? op : old`, since assuming `old` (the value in memory aka "the atomic objects value") is the "first argument", `std::min` requires the first argument (`old`) to be returned when the arguments compare equivalent.
- For a similar reason, `fetch_max(op)` should be: `old < op? op : old`.

@miscco @bernhardmgruber may know the details.

Reply:
I am slightly concerned about changing runtime behavior, but on the other hand that is the more correct one

Reply:
I've modified the behavior to match min/max I believe.